mockgcp/mocklogging/logmetric.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/golang/protobuf/ptypes/empty"
+)
+
+type metricsService struct {
+	*MockService
+	pb.UnimplementedMetricsServiceV2Server
+}
+
+func (s *metricsService) GetLogMetric(ctx context.Context, req *pb.GetLogMetricRequest) (*pb.LogMetric, error) {
+	name, err := s.parseLogMetricName(req.MetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.LogMetric{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "metric %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading metric: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *metricsService) CreateLogMetric(ctx context.Context, req *pb.CreateLogMetricRequest) (*pb.LogMetric, error) {
+	reqName := req.Parent + "/metrics/" + req.GetMetric().GetName()
+	name, err := s.parseLogMetricName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	now := time.Now()
+
+	obj := proto.Clone(req.GetMetric()).(*pb.LogMetric)
+	obj.Name = name.MetricName
+	obj.CreateTime = timestamppb.New(now)
+	obj.UpdateTime = timestamppb.New(now)
+
+	// Not clear why, but we return the values before populating the MetricDescriptor Name/Type
+	retval := proto.Clone(obj).(*pb.LogMetric)
+
+	s.populateDefaultsForLogMetric(name, obj)
+
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating metric: %v", err)
+	}
+
+	return retval, nil
+}
+
+func (s *metricsService) populateDefaultsForLogMetric(name *logMetricName, obj *pb.LogMetric) {
+
+	if obj.MetricDescriptor != nil {
+		obj.MetricDescriptor.Name = fmt.Sprintf("projects/%s/metricDescriptors/logging.googleapis.com/user/%s", name.Project.ID, name.MetricName)
+		obj.MetricDescriptor.Type = fmt.Sprintf("logging.googleapis.com/user/%s", name.MetricName)
+	}
+}
+
+func (s *metricsService) UpdateLogMetric(ctx context.Context, req *pb.UpdateLogMetricRequest) (*pb.LogMetric, error) {
+	reqName := req.MetricName
+
+	name, err := s.parseLogMetricName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+	existing := &pb.LogMetric{}
+	if err := s.storage.Get(ctx, fqn, existing); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "metric %q not found", reqName)
+		}
+		return nil, status.Errorf(codes.Internal, "error reading metric: %v", err)
+	}
+
+	now := time.Now()
+
+	updated := req.GetMetric()
+	updated.Name = name.MetricName
+	updated.CreateTime = existing.CreateTime
+	updated.UpdateTime = timestamppb.New(now)
+
+	// Not clear why, but we return the values before populating the MetricDescriptor Name/Type
+	retval := proto.Clone(updated).(*pb.LogMetric)
+
+	s.populateDefaultsForLogMetric(name, updated)
+
+	if err := s.storage.Update(ctx, fqn, updated); err != nil {
+		return nil, status.Errorf(codes.Internal, "error updating logmetric: %v", err)
```

Comment:
I believe even Update can return a not found error which should be returned as such and not an internal.

Reply:
I mean we technically should be running this in a transaction (or with optimistic concurrency), but we just fetched the object and got it.  I guess it's technically possible, but I'd rather figure out how to get OCC going.